libs/json/encoder.go

```diff
@@ -0,0 +1,246 @@
+
+	case reflect.Array, reflect.Slice:
+		return encodeReflectList(w, rv)
+
+	case reflect.Map:
+		return encodeReflectMap(w, rv)
+
+	case reflect.Struct:
+		return encodeReflectStruct(w, rv)
+
+	// 64-bit integers are emitted as strings, to avoid precision problems with e.g.
+	// Javascript which uses 64-bit floats (having 53-bit precision).
+	case reflect.Int64, reflect.Int:
+		return writeStr(w, `"`+strconv.FormatInt(rv.Int(), 10)+`"`)
+
+	case reflect.Uint64, reflect.Uint:
+		return writeStr(w, `"`+strconv.FormatUint(rv.Uint(), 10)+`"`)
+
+	// For everything else, defer to the stdlib encoding/json encoder
+	default:
+		return encodeStdlib(w, rv.Interface())
+	}
+}
+
+func encodeReflectList(w io.Writer, rv reflect.Value) error {
+	// Emit nil slices as null.
+	if rv.Kind() == reflect.Slice && rv.IsNil() {
+		return writeStr(w, `null`)
+	}
+
+	// Encode byte slices as base64 with the stdlib encoder.
+	if rv.Type().Elem().Kind() == reflect.Uint8 {
+		return encodeStdlib(w, rv.Interface())
+	}
+
+	// Anything else we recursively encode ourselves.
+	length := rv.Len()
+	if err := writeStr(w, `[`); err != nil {
+		return err
+	}
+	for i := 0; i < length; i++ {
+		if err := encodeReflect(w, rv.Index(i)); err != nil {
+			return err
+		}
+		if i < length-1 {
+			if err := writeStr(w, `,`); err != nil {
+				return err
+			}
+		}
+	}
+	return writeStr(w, `]`)
+}
+
+func encodeReflectMap(w io.Writer, rv reflect.Value) error {
+	if rv.Type().Key().Kind() != reflect.String {
+		return errors.New("map key must be string")
+	}
+
+	if err := writeStr(w, `{`); err != nil {
+		return err
+	}
+	writeComma := false
+	for _, keyrv := range rv.MapKeys() {
+		if writeComma {
+			if err := writeStr(w, `,`); err != nil {
+				return err
+			}
+		}
+		if err := encodeStdlib(w, keyrv.Interface()); err != nil {
+			return err
+		}
+		if err := writeStr(w, `:`); err != nil {
+			return err
+		}
+		if err := encodeReflect(w, rv.MapIndex(keyrv)); err != nil {
+			return err
+		}
+		writeComma = true
+	}
+	return writeStr(w, `}`)
+}
+
+func encodeReflectStruct(w io.Writer, rv reflect.Value) error {
+	sInfo := makeStructInfo(rv.Type())
+	if err := writeStr(w, `{`); err != nil {
+		return err
+	}
+	writeComma := false
+	for i, fInfo := range sInfo.fields {
+		frv := rv.Field(i)
+		if fInfo.hidden || (fInfo.omitEmpty && frv.IsZero()) {
+			continue
+		}
+
+		if writeComma {
+			if err := writeStr(w, `,`); err != nil {
+				return err
+			}
+		}
+		if err := encodeStdlib(w, fInfo.jsonName); err != nil {
+			return err
+		}
+		if err := writeStr(w, `:`); err != nil {
+			return err
+		}
+		if err := encodeReflect(w, frv); err != nil {
+			return err
+		}
+		writeComma = true
+	}
+	return writeStr(w, `}`)
+}
+
+func encodeReflectInterface(w io.Writer, rv reflect.Value) error {
+	// Get concrete value and dereference pointers.
+	for rv.Kind() == reflect.Ptr || rv.Kind() == reflect.Interface {
+		if rv.IsNil() {
+			return writeStr(w, `null`)
```

Comment:
Nit: Ideally, all cases where `null` are written should be explicitly documented in the package documentation.

Reply:
This really depends on what we're encoding though, it's up to the caller whether an input is a pointer or a bare value (the first will emit `null`, the latter will emit the zero value).